nc/nc.go

```diff
@@ -287,30 +370,74 @@ func (c *Client) receiveMessage(_ mqtt.Client, msg mqtt.Message) {
 	}
 }
 
-func (c *Client) sendDeliveryConfirmation(messageID string) error {
-	payload, err := json.Marshal(ConfirmationPayload{
-		MessageID: messageID,
-	})
-	if err != nil {
-		return fmt.Errorf("marshaling confirmation payload: %w", err)
-	}
-	if token := c.client.Publish(topicDelivered, 1, false, payload); token.Wait() && token.Error() != nil {
-		return token.Error()
+func (c *Client) ncClientManagementLoop(ctx context.Context) error {
+	managementChan := make(chan interface{})
+
+	log.Println(logPrefix + " starting management loop")
+
+	var client mqtt.Client
+
+	credentialsInvalidated := false
+	credentials, err := c.credsFetcher.GetCredentialsFromConfig()
+	if err == ErrInvalidCredentials {
+		// Client will be initialized when connecting if credentials are invalidated. We want to do this as a part of
+		// connection loop, because we might not have internet connection at this point
+		credentialsInvalidated = true
+	} else if err != nil {
+		return fmt.Errorf("fetching credentials: %w", err)
+	} else {
+		opts := c.getClientOptions(credentials, managementChan, ctx)
+		client = c.clientBuilder.Build(opts)
 	}
+
+	go func() {
+		client = c.connectWithBackoff(client, credentialsInvalidated, managementChan, ctx)
+
+		for {
+			select {
+			case <-ctx.Done():
+				log.Println(logPrefix + " stopping management loop")
+				if client != nil {
+					client.Unsubscribe(unsubscriptions...)
+					client.Disconnect(0)
+					client = nil
+				}
+				return
+			case event := <-managementChan:
+				switch ev := event.(type) {
+				case authLost:
+					client = c.connectWithBackoff(client, true, managementChan, ctx)
+				case connectionLost:
+					client = c.connectWithBackoff(client, false, managementChan, ctx)
+				case mqttMessage:
+					c.handleMessage(client, ev.message)
+				}
+			}
+		}
+	}()
+
 	return nil
 }
 
-func (c *Client) sendAcknowledgement(messageID, trackType, actionSlug string) error {
-	payload, err := json.Marshal(AcknowledgementPayload{
-		MessageID:  messageID,
-		TrackType:  trackType,
-		ActionSlug: actionSlug,
-	})
-	if err != nil {
-		return fmt.Errorf("marshaling acknowledgement payload: %w", err)
-	}
-	if token := c.client.Publish(topicAcknowledged, 1, false, payload); token.Wait() && token.Error() != nil {
-		return token.Error()
+// Start initiates the connection with the NC server and subscribes to mandatory topics
+func (c *Client) Start() error {
+	log.Println(logPrefix + " start")
+
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	c.cancelConnecting = cancelFunc
+
+	if err := c.ncClientManagementLoop(ctx); err != nil {
+		return fmt.Errorf("starting NC management loop: %w", err)
 	}
+
+	return nil
+}
+
+// Stop unsubscribes the topics and drops a connection with the NC server
+func (c *Client) Stop() error {
+	log.Println(logPrefix + " stop")
+
+	c.cancelConnecting()
```

Comment:
This will panic if called before `Start`. Add either status or at least a nil check

Reply:
Fixed, GH didn't mark comment as outdated.